src/index/txindex.cpp

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <index/txindex.h>
+#include <init.h>
+#include <tinyformat.h>
+#include <ui_interface.h>
+#include <util.h>
+#include <validation.h>
+#include <warnings.h>
+
+template<typename... Args>
+static void FatalError(const char* fmt, const Args&... args)
+{
+    std::string strMessage = tfm::format(fmt, args...);
+    SetMiscWarning(strMessage);
+    LogPrintf("*** %s\n", strMessage);
+    uiInterface.ThreadSafeMessageBox(
+        "Error: A fatal internal error occurred, see debug.log for details",
+        "", CClientUIInterface::MSG_ERROR);
+    StartShutdown();
+}
+
+TxIndex::TxIndex(std::unique_ptr<TxIndexDB> db) :
+    m_db(std::move(db)), m_synced(false), m_best_block_index(nullptr)
+{}
+
+bool TxIndex::Init()
+{
+    LOCK(cs_main);
+
+    const CBlockIndex* chain_tip = chainActive.Tip();
+    uint256 tip_hash;
+    if (chain_tip) {
+        tip_hash = chain_tip->GetBlockHash();
+    }
+
+    // Attempt to migrate txindex from the old database to the new one. Even if
+    // chain_tip is null, the node could be reindexing and we still want to
+    // delete txindex records in the old database.
+    if (!m_db->MigrateData(*pblocktree, tip_hash)) {
+        return false;
+    }
+
+    if (!chain_tip) {
+        m_synced = true;
+        return true;
+    }
+
+    uint256 best_block_hash;
+    if (!m_db->ReadBestBlockHash(best_block_hash)) {
+        FatalError("%s: Failed to read from tx index database", __func__);
+        return false;
+    }
+
+    if (best_block_hash.IsNull()) {
+        return true;
+    }
+
+    const CBlockIndex* pindex = LookupBlockIndex(best_block_hash);
+    if (!pindex) {
+        FatalError("%s: Last block synced by txindex is unknown", __func__);
```

Comment:
I don't understand. This line is in the `Init` method, not the BlockConnected callback. As long as we never prune entries from the block index, I don't see when this check would fail.

Reply:
A block may be connected, but the block index may not be written to disk until much later. Thus, we can run far ahead here.

Reply:
See last two commits (c837c6c & 297f89c) for fix.